activesupport/lib/active_support/error_reporter.rb

```diff
@@ -0,0 +1,97 @@
+  #
+  #   Rails.error.handle do
+  #     do_something!
+  #   end
+  #
+  # If an error is raised, it will be reported and swallowed.
+  #
+  # Alternatively if you want to report the error but not swallow it, you can use `record`
+  #
+  #   Rails.error.record do
+  #     do_something!
+  #   end
+  #
+  # Both methods can be restricted to only handle a specific exception class
+  #
+  #  maybe_tags = Rails.error.handle(Redis::BaseError) { redis.get("tags) }
+  #
+  # You can also pass some extra context information that may be used by the error subscribers:
+  #
+  #  Rails.error.handle(context: { section: "admin" }) do
+  #    # ...
+  #  end
+  #
+  # Additionally a +severity+ can be passed along to communicate how important the error report is.
+  # +severity+ can be one of +:error+, +:warning+ or +:info+. Handled errors default to the +:warning+
+  # severity, and unhandled ones to +error+.
+  class ErrorReporter
+    SEVERITIES = %i(error warning info)
+
+    def initialize(*subscribers)
+      @subscribers = subscribers.flatten
+    end
+
+    # Report any unhandled exception, and swallow it.
+    #
+    #  Rails.error.handle do
+    #    1 + '1'
+    #  end
+    #
+    def handle(error_class = StandardError, severity: :warning, context: {})
+      yield
+    rescue error_class => error
+      report(error, handled: true, severity: severity, context: context)
+      nil
+    end
+
+    def record(error_class = StandardError, severity: :error, context: {})
+      yield
+    rescue error_class => error
+      report(error, handled: false, severity: severity, context: context)
+      raise
+    end
+
+    # Register a new error subscriber. The subscriber must respond to
+    #
+    #   report(Exception, handled: Boolean, context: Hash)
+    #
+    # The +report+ method +should+ never raise an error.
+    def subscribe(subscriber)
+      unless subscriber.respond_to?(:report)
+        raise ArgumentError, "Error subscribers must respond to #report"
+      end
+      @subscribers << subscriber
+    end
+
+    # Update the execution context that is accessible to error subscribers
+    #
+    #   Rails.error.set_context(section: "checkout", user_id: @user.id)
+    #
+    # See +ActiveSupport::ExecutionContext.set+
+    def set_context(...)
+      ActiveSupport::ExecutionContext.set(...)
+    end
+
+    # When the block based +handle+ and +record+ methods are not suitable, you can directly use +report+
+    #
+    #  Rails.error.report(error, handled: true)
+    def report(error, handled:, severity: handled ? :warning : :error, context: {})
+      unless SEVERITIES.include?(severity)
+        raise ArgumentError, "severity must be one of #{SEVERITIES.map(&:inspect).join(", ")}, got: #{severity.inspect}"
+      end
+
+      full_context = ActiveSupport::ExecutionContext.to_h.merge(context)
+      @subscribers.each do |subscriber|
+        subscriber.report(error, handled: handled, severity: severity, context: full_context)
```

Comment:
if one subscriber raises an error during reporting, will it interrupts the reporters after it?

Reply:
I think the `subscriber.report` call should itself be wrapped in a `rescue` that falls back to the logs if the subscriber throws an error. The subscriber _shouldn't_, but we can't rely on them not to.

Reply:
Yeah, subscribers are expected to never raise, but we definitely shouldn't blindly trust them.

Reply:
Hum, so the difficulty here is to have access to a logger. I'll have to dig more into this.